consensus/polybft/state_sync_manager.go

```diff
@@ -0,0 +1,469 @@
+		AggSignature: aggregatedSignature,
+		PublicKeys:   publicKeys,
+	}
+
+	return msg, nil
+}
+
+// getAggSignatureForCommitmentMessage checks if pending commitment has quorum,
+// and if it does, aggregates the signatures
+func (s *StateSyncManager) getAggSignatureForCommitmentMessage(commitment *Commitment) (Signature, [][]byte, error) {
+	validatorSet := s.validatorSet
+
+	validatorAddrToIndex := make(map[string]int, validatorSet.Len())
+	validatorsMetadata := validatorSet.Accounts()
+
+	for i, validator := range validatorsMetadata {
+		validatorAddrToIndex[validator.Address.String()] = i
+	}
+
+	commitmentHash, err := commitment.Hash()
+	if err != nil {
+		return Signature{}, nil, err
+	}
+
+	// get all the votes from the database for this commitment
+	votes, err := s.state.getMessageVotes(commitment.Epoch, commitmentHash.Bytes())
+	if err != nil {
+		return Signature{}, nil, err
+	}
+
+	var signatures bls.Signatures
+
+	publicKeys := make([][]byte, 0)
+	bitmap := bitmap.Bitmap{}
+	signers := make(map[types.Address]struct{}, 0)
+
+	for _, vote := range votes {
+		index, exists := validatorAddrToIndex[vote.From]
+		if !exists {
+			continue // don't count this vote, because it does not belong to validator
+		}
+
+		signature, err := bls.UnmarshalSignature(vote.Signature)
+		if err != nil {
+			return Signature{}, nil, err
+		}
+
+		bitmap.Set(uint64(index))
+
+		signatures = append(signatures, signature)
+		publicKeys = append(publicKeys, validatorsMetadata[index].BlsKey.Marshal())
+		signers[types.StringToAddress(vote.From)] = struct{}{}
+	}
+
+	if !validatorSet.HasQuorum(signers) {
+		return Signature{}, nil, errQuorumNotReached
+	}
+
+	aggregatedSignature, err := signatures.Aggregate().Marshal()
+	if err != nil {
+		return Signature{}, nil, err
+	}
+
+	result := Signature{
+		AggregatedSignature: aggregatedSignature,
+		Bitmap:              bitmap,
+	}
+
+	return result, publicKeys, nil
+}
+
+type PostEpochRequest struct {
+	// BlockNumber is the number of the block being executed
+	BlockNumber uint64
+
+	// NewEpochID is the id of the new epoch
+	NewEpochID uint64
+
+	// SystemState is the state of the governance smart contracts
+	// after this block
+	SystemState SystemState
+
+	// ValidatorSet is the validator set for the new epoch
+	ValidatorSet *validatorSet
+}
+
+// PostEpoch notifies the state sync manager that an epoch has changed,
+// so that it can discard any previous epoch commitments, and build a new one (since validator set changed)
+func (s *StateSyncManager) PostEpoch(req *PostEpochRequest) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// build a new commitment at the end of the epoch
+	nextCommittedIndex, err := req.SystemState.GetNextCommittedIndex()
+	if err != nil {
+		return err
+	}
+
+	commitment, err := s.buildCommitment(req.NewEpochID, nextCommittedIndex)
+	if err != nil {
+		return err
+	}
+
+	s.commitment = commitment
+	s.validatorSet = req.ValidatorSet
+	s.epoch = req.NewEpochID
```

Comment:
Isn't it enough just to add `Lock` for these two lines (if we are going to switch to `RWMutex`)? Basically we need lock just because messages are arriving from different goroutines?

Reply:
RW lock will be added through another PR, for multiple commitments.